### ISSUE TITLE
How can I implement adaptation algorithms that have access to information about the last downloaded (DASH) segment/chunk?

### ISSUE THREAD

**Issue by emdash-ie** [google/ExoPlayer#3053]
I would like to implement custom (DASH) video adaptation algorithms in ExoPlayer, where the algorithms have access to information about each segment downloaded (mainly the size, the position in the stream, and the time duration of the segments), on which they can base their choice.

Starting from the demo application, I have tried to implement a new `TrackSelection` (and Factory) following the example of `AdaptiveTrackSelection`. I am having difficulty getting the required information about the segments to the new `TrackSelection` implementation.

At first, I tried registering an `AdaptiveMediaSourceEventListener` on the media sources (replacing the `EventLogger` in the demo app), which the `TrackSelection` could then grab the data from. This works, but the data arrives at the `AdaptiveMediaSourceEventListener` too slowly, and so sometimes the `TrackSelection` is reusing information from the last segment when it is called.

Then, I found [this issue](https://github.com/google/ExoPlayer/issues/2909), where it says that `AdaptiveMediaSourceEventListener` is only intended for logging, not for changing the behaviour of the player, which makes sense. I possibly should have realised this [when I was told about the playback thread versus the background thread](https://github.com/google/ExoPlayer/issues/3042) – sorry about that!

I can see that the information I'm looking for is available inside `DefaultDashChunkSource.getNextChunk()` as the `previous` parameter, as well as in `ChunkSampleStream`, but I'm unsure how to get it from there to my `TrackSelection`.

Could I get some advice for how I should aim to go about this? Will I have to reimplement `DashChunkSource` as well as `TrackSelection`?

Thanks for all the help so far :)

**Comment by zhanghuicuc** [comment]
maybe using selectionData
see https://github.com/zhanghuicuc/SimplePlayer/blob/dev-exo/library/src/main/java/com/google/android/exoplayer2/trackselection/QAdaptTrackSelection.java

**Comment by emdash-ie** [comment]
Thank you for your reply. I see that you had to modify some of the existing ExoPlayer library to achieve what you wanted – I was hoping to avoid this. Do you know if that's possible?

I'm also not sure if selectionData really helps me here, as I'm looking to pass information into the TrackSelection, rather than get information from it – am I missing something?

**Comment by soadri** [comment]
If you want to pass the information, I think that using listener is best

**Comment by emdash-ie** [comment]
Thanks for the suggestion – is there a particular listener you'd recommend?

**Comment by soadri** [comment]
What is your adaptation concept?

If the concept is based on the bandwidth, the listener in the BandwidthMeter is good

**Comment by emdash-ie** [comment]
Sorry for the delay in getting back to you!

I want to be able to provide the algorithms with the following information for each chunk:

* Chunk index (or chunk start/end time)
* Chunk arrival time (the time the chunk finished downloading)
* Chunk download duration
* Total stall duration (how much time the player has spent stalled)
* Track bitrate
* Chunk size (in bits)
* Chunk length (in seconds)
* Current buffer level

At first, I went through a bit of a laborious process to get these values: I forked the ExoPlayer library and replaced `ChunkSampleStream` and `DefaultDashChunkSource` with my own custom classes, which passed this information to my `TrackSelection` subtype. In order to use my replacement for `ChunkSampleStream`, I then had to replace `DashMediaPeriod` and `DashMediaSource` as well.

After this I realised that I can get the information I need by implementing just a `DashChunkSource` and `TransferListener` (replacing the `DefaultBandwidthMeter`, as you pointed out!), as I'm working with video-only streams. This means that I can import the ExoPlayer library instead of forking it.

For the moment, this is my solution, though if I'm working with streams that have audio, I'll need another way to get the arrival time and load duration of each chunk.
